autotools: update to latest (2022)

mcprat opened the pull request:
Libtool being handled in #4595
(but now merged here)

should be tested with SDK to see if there is a similar problem as the libtool situation

cotequeiroz committed:
tools/libtool: bump to 2.4.6

This updates libtool to its current release, from 2015.  Current patches
were renumbered and given a description text.  The fix in
160-passthrough-ssp.patch is no longer needed.

A patch to speed up build was cherry-picked, and another openwrt
specific patch was needed to not use quotes in $(SHELL), to acommodate
our "SHELL=/usr/bin/env bash" usage.

The already present call to ./bootstrap ensures that generated files are
refreshed, so the patches are applied only to their sources.  Also, that
bootstrap call was adjusted to run at the appropriate time when QUILT=1.

References below are relevant commits to upstream libtool
regarding some of the changes to patches.

This commit is being reapplied after previous revertion, and after some editing.
The fix for the issue that prompted reverting is the parent of this commit.

Ref: 435cb8d71 ("libtoolize: simplify runtime by substituting pkgauxdir")
Ref: 3cf11cfe2 ("libtoolize: rewritten over funclib.sh instead of general.m4sh")
Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>
Signed-off-by: Sebastian Kemper <sebastian_ml@gmx.net>
[refactored to simplify patch changes, expanded patches, added upstream references]
Signed-off-by: Michael Pratt <mcpratt@pm.me>

micmac1 committed:
tools/libtool: fix hard-coded paths

libtoolize hardcodes some paths. This is fine when building libtool and
then using it (for example in OpenWrt's buildroot). But when using an
SDK the paths are most likely different.

For example, when building util-linux within an SDK we're greeted with
the following message:

  libtoolize: error: $pkgauxdir is not a directory: '/path/to/openwrt/staging_dir/host/share/libtool/build-aux'

This is because staging_dir/host/bin/libtoolize contains the following
hard-coded paths from when the SDK was built in the first place:

  prefix="/path/to/openwrt/staging_dir/host"
  datadir="/path/to/openwrt/staging_dir/host/share"
  pkgauxdir="/path/to/openwrt/staging_dir/host/share/libtool/build-aux"
  pkgltdldir="/path/to/openwrt/staging_dir/host/share/libtool"
  aclocaldir="/path/to/openwrt/staging_dir/host/share/aclocal"

This commits updates 000-relocatable.patch to correct the paths,
relative to "$STAGING_DIR_HOST".

Ref: 96e05e2e3 ("libtool: Revert "libtool: bump to 2.4.6"")
Signed-off-by: Sebastian Kemper <sebastian_ml@gmx.net>
[adapted to older libtool version before bumping, use STAGING_DIR_HOST]
Signed-off-by: Michael Pratt <mcpratt@pm.me>

neheb committed:
tools/libtool: bump to 2.4.7

Remove upstreamed patch and refresh others.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
[rebased and refreshed patches again]
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/autoconf: bump to 2.71

Autoconf release 2.71 is a bugfix release of 2.70
Links provided are the changelog of 2.70 [1] [2]
and the announcement of 2.71 [3]

The first patch needed to be rewritten.
Some of the lines in the patch were moved
to be consistent with the same change added upstream
in other similar files.

Second and third patches are no longer needed,
Emacs can now be disabled at the configure stage,
and support for musl was added upstream.

The patch now causes help2man to be required
for installing man pages, but we don't need them.
There is no way to disable man pages build
with the configure script,
so use make to touch the files with the build target install-man1.

Link: https://lists.gnu.org/archive/html/autotools-announce/2020-12/msg00001.html
Link: https://lwn.net/Articles/839395
Link: https://lists.gnu.org/archive/html/info-gnu/2021-01/msg00017.html
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/autoconf-archive: bump to 2022.09.03

Changelog included in package.
Link to announcement.

Link: https://lists.gnu.org/archive/html/info-gnu/2022-09/msg00004.html
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/automake: bump to 1.16.5

Automake releases 1.16.2 to 1.16.5 are major releases,
the two before are minor releases.

The first patch needed rewriting, some lines moved
to be consistent with the same change upstream in other files.
This also simplifies the patch.

Second patch refreshed.

Links to changelogs in order, oldest to newest.

Link: https://lists.gnu.org/archive/html/info-gnu/2018-02/msg00008.html
Link: https://lists.gnu.org/archive/html/info-gnu/2018-03/msg00002.html
Link: https://lists.gnu.org/archive/html/info-gnu/2020-03/msg00009.html
Link: https://lists.gnu.org/archive/html/info-gnu/2020-11/msg00005.html
Link: https://lists.gnu.org/archive/html/info-gnu/2021-07/msg00013.html
Link: https://lists.gnu.org/archive/html/info-gnu/2021-10/msg00000.html
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
works fine on my system,

error in CI is this

looking into how to disable man pages build

```
Updating man page man/autoconf.1
/usr/bin/mkdir -p man
PATH="./man:$PATH"; \
PERL="/usr/bin/env perl"; \
PACKAGE_NAME="GNU Autoconf"; \
VERSION="2.71"; \
RELEASE_YEAR="2021"; \
top_srcdir="."; \
channeldefs_pm="lib/Autom4te/ChannelDefs.pm"; \
export PATH PERL PACKAGE_NAME VERSION RELEASE_YEAR; \
export top_srcdir channeldefs_pm; \
/usr/bin/env bash '/home/runner/work/openwrt/openwrt/openwrt/build_dir/host/autoconf-2.71/build-aux/missing' help2man \
    --include=./man/autoconf.x \
    --include=./man/common.x \
    --source='GNU Autoconf 2.71' \
    --output=man/autoconf.1.t autoconf.w
/home/runner/work/openwrt/openwrt/openwrt/build_dir/host/autoconf-2.71/build-aux/missing: line 81: help2man: command not found
WARNING: 'help2man' is missing on your system.
         You should only need it if you modified a dependency of a man page.
         You may want to install the GNU Help2man package:
         <https://www.gnu.org/software/help2man/>
make[4]: *** [Makefile:2216: man/autoconf.1] Error 127
make[4]: Leaving directory '/home/runner/work/openwrt/openwrt/openwrt/build_dir/host/autoconf-2.71'
make[3]: *** [Makefile:928: all] Error 2
make[3]: Leaving directory '/home/runner/work/openwrt/openwrt/openwrt/build_dir/host/autoconf-2.71'
make[2]: *** [Makefile:39: /home/runner/work/openwrt/openwrt/openwrt/build_dir/host/autoconf-2.71/.built] Error 2
make[2]: Leaving directory '/home/runner/work/openwrt/openwrt/openwrt/tools/autoconf'
time: tools/autoconf/compile#1.11#0.38#3.17
```

mpratt14 commented:
ok so now, just needs testing with all (or most) packages and with SDK

mpratt14 commented:
added backport patches for autoconf-archive and gettext

https://github.com/autoconf-archive/autoconf-archive/commit/c92672b9c59f9011e5d68220e99cfdd70c1b98df
https://github.com/autoconf-archive/autoconf-archive/commit/13018ae91007cfa1b1a3877db8ef5cb233af1ae6

https://git.savannah.gnu.org/gitweb/?p=gettext.git;a=commit;h=d1836dbbd6a90b4c0ab79bc5292c023f08b49511

mcprat committed:
build: disable automake dependency tracking

Recent versions of Automake
have changed dependency tracking significantly
(reference commit below)
causing breakage in some package builds
when using newer Automake with packages that need autoreconf
that were bootstrapped with an old version of Automake.

Those changes cause a great inconsistency between packages over time
where some packages may or may not use this feature,
and may or may not update the .ac and .am files
to work with the new methods.
This problem might exist in many packages
where autoreconf is not currently required,
but would cause build failure if autoreconf is used.

Fortunately, this feature is practically useless
for the purposes of Openwrt and the average developer,
so we can disable it.

GNU Automake manual states in part:

  "Because dependencies are only computed as a side-effect of compilation...
   no dependency information exists the first time a package is built...
   dependency tracking is completely useless for one-time builds..."

A nice side-effect is that build times are slightly faster.

Ref: 6a675ef17edf7109da189f5ae70e2dc6b7665896 (automake.git)
Signed-off-by: Michael Pratt <mcpratt@pm.me>

PolynomialDivision commented:
Already built an image for a Linksys E8450. Works as expected. :)
I pushed a workflow to build everything (I hope it is correct...) https://github.com/PolynomialDivision/openwrt/runs/8085907974?check_suite_focus=true

Nice to see autotools is finally updated again. :D

mpratt14 commented:
*almost* ready but not quite yet

tools and targets compile ok
almost done testing all packages

There is (so far) one common problem among a few packages, which is dependency tracking, and I'm going to push a commit for that here soon (to disable it since its not very useful)

some reading from GNU docs
https://www.gnu.org/software/automake/manual/html_node/Dependency-Tracking.html

mpratt14 commented:
also, I'm going to rename the patch for gettext, I think backports should be first in the list...

mpratt14 commented:
should I be adding the symlink between `aclocal-1.16` and `aclocal`?

also I'd like to learn the reason we need all those symlinks

Vladdrako commented:
libnftnl doesn't compile
<pre>
make[5]: Entering directory '/home/vladdrako/openwrt/build_dir/target-mipsel_24kc_musl/libnftnl-1.2.3/src'
 cd .. && /bin/sh /home/vladdrako/openwrt/build_dir/target-mipsel_24kc_musl/libnftnl-1.2.3/build-aux/missing automake-1.16 --foreign src/Makefile
configure.ac:10: error: version mismatch.  This is Automake 1.16.5,
configure.ac:10: but the definition used by this AM_INIT_AUTOMAKE
configure.ac:10: comes from Automake 1.16.3.  You should recreate
configure.ac:10: aclocal.m4 with aclocal and run automake again.
WARNING: 'automake-1.16' is probably too old.
         You should only need it if you modified 'Makefile.am' or
         'configure.ac' or m4 files included by 'configure.ac'.
         The 'automake' program is part of the GNU Automake package:
         <https://www.gnu.org/software/automake>
         It also requires GNU Autoconf, GNU m4 and Perl in order to run:
         <https://www.gnu.org/software/autoconf>
         <https://www.gnu.org/software/m4/>
         <https://www.perl.org/>
make[5]: *** [Makefile:479: Makefile.in] Error 63
</pre>

PolynomialDivision commented:
> libnftnl doesn't compile

On my CI it compiled fine. Can you share `./scripts/diffconfig.sh`?
See here:
https://github.com/PolynomialDivision/openwrt/runs/8085907974?check_suite_focus=true

Vladdrako commented:
@PolynomialDivision Oh, I found the issue. It was the fullcone patch. Sorry for the noise :)

mcprat committed:
tools/automake: use make variable to generate aclocal symlinks

Instead of having a somewhat random list of aclocal symlinks
which are named with the API versioning scheme,
install a symlink for every API version since 1.11
(the first release after Automake was moved to git)
using the API version number from the version of automake
that is currently in openwrt.

Automake API versioning does not include the patch level number.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
added commit to improve the aclocal symlinks

@PolynomialDivision I finished testing with all packages on SDK, packages that failed were for other reasons, the rest handled in the PR I made in the packages repo.

I think this is ready now

PolynomialDivision commented:
> also I'd like to learn the reason we need all those symlinks

Did you find out the reason for the symlinks?

mpratt14 commented:
not exactly, git history here doesn't explain anything about it. I am guessing that it's possible that packages want a specific aclocal version even when autoreconf is used, and the symlinks prevent a host machine's version of aclocal being used over openwrt's version.

PolynomialDivision commented on tools/automake/Makefile:
```diff
@@ -36,11 +37,13 @@ define Host/Install
-	ln -sf aclocal $(STAGING_DIR_HOST)/bin/aclocal-1.15
+	( \
+		api=$(PKG_API_VERSION); \
+		while [ "$$$$api" -ne 10 ]; do \
```
Why `-ne` and not `-le`?

mpratt14 commented:
autoconf-archive is preparing a new release (possibly because I bugged them about it hehe)

so waiting a little bit, then we wont need those 2 patches

mpratt14 commented:
testing with CI a better way to disable man pages build

mcprat committed:
tools/autoconf: clean up build recipes

The configure option datarootdir
defaults to PREFIX/share.

The Host/Clean define should be Host/Uninstall
otherwise it is removing the build directory
before there is a chance to try "make uninstall"
in that directory.
Host/Clean/Default is no longer defined.

Use default build recipes defined in host-build.mk
instead of custom ones that are equivalent.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/autoconf-archive: clean up build recipes

The configure option datarootdir
defaults to PREFIX/share.

The Host/Clean define should be Host/Uninstall
otherwise it is removing the build directory
before there is a chance to try "make uninstall"
in that directory.
Host/Clean/Default is no longer defined.

Use default build recipes defined in host-build.mk
instead of custom ones that are equivalent.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/automake: clean up build recipes

The configure option datarootdir
defaults to PREFIX/share.

The Host/Clean define should be Host/Uninstall
otherwise it is removing the build directory
before there is a chance to try "make uninstall"
in that directory.
Host/Clean/Default is no longer defined.

Consolidate the install stage with Host/Uninstall,
since it is essentially uninstalling before installing.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@PolynomialDivision ready for review again 👍🏼

PolynomialDivision commented:
Just added again a workflow for testing the packages of the base repository:
https://github.com/PolynomialDivision/openwrt/runs/8238339272?check_suite_focus=true

For me everything looks fine. Great work. :)
However, I have no merge rights. That's why I think a review from @hauke would be nice.

PolynomialDivision commented:
Somehow `kernel/bpf-headers` failed to compile. Just compiling on my local pc again. However, this will take sometime... ^^

hauke commented on tools/autoconf/Makefile:
```diff
@@ -7,18 +7,20 @@
 
 define Host/Compile
+	-$(call Host/Compile/Default,--touch install-man1) # disable building of man pages
 	$(call Host/Compile/Default)
```
Why do you call `Host/Compile/Default` twice?

mpratt14 replied:
see first comment by me. CI build fails to build manual, so we just skip it by touching the files

hauke commented:
For me the `--disable-dependency-tracking` looks dangerous. Will automake always recompile the complete application?

mpratt14 commented:
all dependency tracking does is try to figure out whether there are things make should *not* do, like the GNU manual says, if youre building something for the first time, there is no effect

mpratt14 commented:
when the build system decides to recompile something, the whole build directory for that package is deleted and replaced first (beginning of STAMP_PREPARED rules) so whatever dependency tracking info is deleted anyway

I have observed that packages that don't use the feature simply print a warning like "option --disable-dependency-tracking not used in this project" or something like that

mpratt14 commented:
@PolynomialDivision do we know why bpf-headers failed? it seems strange, but looks unrelated to autotools

mpratt14 commented:
I had saved the previous state of the PR in another branch, if it works the old way then we know something's wrong here

https://github.com/mpratt14/openwrt/tree/autotools-2022-old

neheb commented:
conflicts

PolynomialDivision commented:
@mpratt14 We updatet gettext-full to 0.21.1 so you can just drop the commit adding this patch " libtextstyle: Make gl_INIT_PACKAGE work with Automake >= 1.16.4."

mpratt14 commented:
right, I was rebasing this but I didn't push because I didn't test build all packages yet

also I would like to include libtool, since the person running that PR gave up on it in #4595

I would like to wait for the SDK fix in #11379 but otherwise I will be getting back to testing this eventually

PolynomialDivision commented:
What about updating and adding libtool to this PR regardless of the SDK fix? Than I would also give it a test-run.

lovehackintosh commented:
autoconf-archive-2023.02.20 released

mcprat committed:
tools/libtool: refresh patches

Preparation for bumping libtool several versions at once,
which includes some complex changes to patches.

Remove an empty line change from a patch.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/libtool: add commit messages to patches

Add commit messages to patches as intended
by the last bump to libtool which had to be reverted.

This allows for a cleaner diff later,
removing this clutter from the complex changes to patches
that are required to bump to the latest libtool version.

Ref: c377d874b ("libtool: bump to 2.4.6")
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
ok, finally rebased this and merged  #4595 however I split up the commits for libtool differently, if it doesn't look right I can always squash them back together. I'm trying to get the cleanest simplest possible diff for the big bump from libtool 2.4.2 to 2.4.6.

ping @cotequeiroz @micmac1 @neheb if you have any comments regarding how I'm handling the re-implementation of your commits for libtool, I would like your opinions.

I'm not going to bother with autoconf-archive 2023.02.20 right now. Let's just get these versions in first which have been somewhat tested already.

the previous state of the PR is in another branch if anyone wanted to look at it
https://github.com/mpratt14/openwrt/tree/autotools-2022-rebase

hauke commented on tools/autoconf/Makefile:
```diff
@@ -15,23 +15,15 @@ PKG_HASH:=64ebcec9f8ac5b2487125a86a7760d2591ac9e1d3dbd59489633f9de62a57684
-
-define Host/Install
-	export SHELL="$(BASH)"; $(MAKE) -C $(HOST_BUILD_DIR) install
+	$(call Host/Compile/Default)
```
`define Host/Compile` is now doing the same as the default one. I think this is not needed any more.

mpratt14 replied:
I add a line there in the commit after, "tools/autoconf: bump to 2.71"

one of the things that changed was how the manual is generated.
that version fails to build on the CI because it doesn't have help2man, so we skip generating the man pages instead.

it's a nicer way to achieve the same result, which I tested sometime last year, but it was ugly and complicated

https://github.com/mpratt14/openwrt/commit/5caa888eebde032500faa54b4e52bd0144416bf8

hauke replied:
Sorry I looked at the cleanup commit 24bf39833b6b745c0f385ca3a8a9c6b686c85e6b and did not saw that you add the extra line some patches later.

dhewg commented:
I've seen it fail to built because of that sometimes. It's rare. Haven't looked at it yet, but at first glance it looks like some parallelism issue? Because the next try always works then?

On March 13, 2023 3:16:43 PM GMT+01:00, Michael Pratt ***@***.***> wrote:
>@dhewg automake is still running for lrzsz, I'm building using the SDK
>
>you have any idea why?
>
>(ref: https://github.com/openwrt/packages/commit/6d6c4b21b5e22a9f1058db5b61521a298e00a5f0)
>

hauke commented:
Thank you for the patches. I applied them to master in 129be247a6a0cb11132691bdf27104f48ab08530, a1527207b8f23d4084a18e988605fd30c5e72ca8, 99ed71faec745ced91a00d46c6512b27c9892d2b, 8d0ff17ec9907b1383644793884c2b50598aafea, a3cf796e7731fb7ce657017ad34e93acf35b038a, b3776230f312972e40ef49ce8b830db0c8138b06, 030447b8f4c70cf8313dd8de5a3f70e2ad1de375, f509935fe3bf85dcd5cce98beb32d426b3ea11f2, c7828a7184d704dafcdfd89b1f59ac2a70090637, efe56a51bcd0b2c2db39c682e88fb166a2431727, a22c54e39f8cb668ea5d717a51359fe2409fa336, 0d0cfe9f2d086404093b547b89afec822c32bd76 and bf261073dc4d788170d51fdc718d63ce0976fd77.

hauke commented:
It looks like this broke something.
gnutls is not building any more, I can easily reproduce it. See the comments here: https://github.com/openwrt/openwrt/commit/030447b8f4c70cf8313dd8de5a3f70e2ad1de375

The SDK looks broken, we get many failures. See the faillog here: https://downloads.openwrt.org/snapshots/faillogs/aarch64_cortex-a72/base/

PolynomialDivision commented:
> gnutls is not building any more, I can easily reproduce it. See the comments here: [030447b](https://github.com/openwrt/openwrt/commit/030447b8f4c70cf8313dd8de5a3f70e2ad1de375)

https://forum.openwrt.org/t/automake-links-to-compile-etc-missing-recent-commit/154588/8

For me it helped to increase the version for the kmod package, e.g.:
https://github.com/openwrt/packages/pull/20672

hauke commented:
It looks like the SDK uses a hard coded path. I useed the SDK from the build bots locally and build this: `make package/libmnl/{clean,compile} V=99`:
```
autoreconf: running: /home/hauke/openwrt/sdk/tmp/work/openwrt-sdk-malta-be_gcc-12.2.0_musl.Linux-x86_64/staging_dir/host/bin/autoconf --include=/home/hauke/openwrt/sdk/tmp/work/openwrt-sdk-malta-be_gcc-12.2.0_musl.Linux-x86_64/staging_dir/target-mips_24kc_musl/host/share/aclocal --include=/home/hauke/openwrt/sdk/tmp/work/openwrt-sdk-malta-be_gcc-12.2.0_musl.Linux-x86_64/staging_dir/hostpkg/share/aclocal --include=/home/hauke/openwrt/sdk/tmp/work/openwrt-sdk-malta-be_gcc-12.2.0_musl.Linux-x86_64/staging_dir/target-mips_24kc_musl/usr/share/aclocal --include=m4 --include=. --prepend-include=/home/hauke/openwrt/sdk/tmp/work/openwrt-sdk-malta-be_gcc-12.2.0_musl.Linux-x86_64/staging_dir/host/share/aclocal --force
autom4te: error: /builder/shared-workdir/build/staging_dir/host/share/autoconf/autoconf/trailer.m4: no such file or directory
autoreconf: error: /home/hauke/openwrt/sdk/tmp/work/openwrt-sdk-malta-be_gcc-12.2.0_musl.Linux-x86_64/staging_dir/host/bin/autoconf failed with exit status: 1
```

autom4te uses the ` /builder/shared-workdir/build` prefix which it should not use.

jow- commented:
The `trailer.m4` related issues should be solved by 0ba130abb2c90c76b6b63f389da02b8bcfe6b78e (requires a `tools/autoconf` and `target/sdk` rebuild).

hauke commented:
The gnutls and kmod problem should get fixed in #12185.

PolynomialDivision commented:
What about
```
configure.ac:13: error: required file 'build-aux/compile' not found
configure.ac:13:   'automake --add-missing' can install 'compile'
```
Are those errors also fixed because configure script was aborted by the error and it never installed `build-aux/compile`?

mpratt14 commented:
sigh, I wasn't quite done testing, but thanks for handling these small regressions

@PolynomialDivision which package you're seeing that?

mpratt14 commented:
BTW some of those faillogs seem unrelated ... and honestly might not be reproducible. I have a hard time believing this broke busybox

```
make[3]: *** No rule to make target '/builder/shared-workdir/build/sdk/build_dir/target-aarch64_cortex-a72_musl/busybox-default/busybox-1.36.0/.prepared_ea9937cc581ae026a40b86cc51935fe7_18f1e190c5d53547fed41a3eaa76e9e9_check', needed by '/builder/shared-workdir/build/sdk/build_dir/target-aarch64_cortex-a72_musl/busybox-default/busybox-1.36.0/.prepared_ea9937cc581ae026a40b86cc51935fe7_18f1e190c5d53547fed41a3eaa76e9e9'.  Stop.
```

mpratt14 commented:
what's the best way to handle gmp? patch `LT_INIT` or try to update it?

```
libtool: Version mismatch error.  This is libtool 2.4.7, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.7
libtool: and run autoconf again.
```

mpratt14 commented:
wait I've seen this before, it needs either autoreconf or autotools templates to not be patched...

hauke commented:
Some build bot runs are looking fine now: https://downloads.openwrt.org/snapshots/faillogs/arm_arm926ej-s/
I think the other failing ones are still using an old SDK.

gnutls failed in the autoreconf step, but OpenWrt ignored the failure and just continues with the old configure script, but they are not working with OpenWrt and then you get the `libtool: and run autoconf again.` error.

If you still see problems please make sure the package was cleaned before and also look into the logs from the autoreconf step. If that failed you can ignore the rest of the log. It would be nice if the package build would fail if autoreconf fails.

Aidener commented:
> libnftnl doesn't compile
> 
> make[5]: Entering directory '/home/vladdrako/openwrt/build_dir/target-mipsel_24kc_musl/libnftnl-1.2.3/src'
>  cd .. && /bin/sh /home/vladdrako/openwrt/build_dir/target-mipsel_24kc_musl/libnftnl-1.2.3/build-aux/missing automake-1.16 --foreign src/Makefile
> configure.ac:10: error: version mismatch.  This is Automake 1.16.5,
> configure.ac:10: but the definition used by this AM_INIT_AUTOMAKE
> configure.ac:10: comes from Automake 1.16.3.  You should recreate
> configure.ac:10: aclocal.m4 with aclocal and run automake again.
> WARNING: 'automake-1.16' is probably too old.
>          You should only need it if you modified 'Makefile.am' or
>          'configure.ac' or m4 files included by 'configure.ac'.
>          The 'automake' program is part of the GNU Automake package:
>          
>          It also requires GNU Autoconf, GNU m4 and Perl in order to run:
>          
>          
>          
> make[5]: *** [Makefile:479: Makefile.in] Error 63

@Vladdrako I get the same error when compiling official master source code for Xiaomi Router4a. I will appreciate that if you can give me some tips about how to repair this error.
